File: leetcode/Distinct_Subsequences.java
import java.util.*;

public class Solution {
    public int numDistinct(String S, String T) {
        int[][] d = new int[S.length() + 1][T.length() + 1];
        for (int i = 0; i <= S.length(); ++i)
            for (int j = 0; j <= T.length(); ++j)
                d[i][j] = 0;
        d[0][0] = 1;

        for (int i = 0; i < S.length(); ++i)
            for (int j = 0; j <= T.length(); ++j)
                if (d[i][j] != 0) {
                    d[i + 1][j] += d[i][j];
                    if (j < T.length() && S.charAt(i) == T.charAt(j))
                        d[i + 1][j + 1] += d[i][j];
                }

        return d[S.length()][T.length()];
    }
}

File: leetcode/Distinct_Subsequences.cpp
class Solution {
public:
    int numDistinct(string S, string T) {
        vector<int> num_matched(T.length() + 1);
        num_matched[0] = 1;
        for (int j = 1; j < num_matched.size(); ++j)
            num_matched[j] = 0;
        for (int i = 0; i < S.length(); ++i)
            for (int j = num_matched.size() - 1; j >= 0; --j)
                if (num_matched[j] > 0 && T[j] == S[i])
                    num_matched[j + 1] += num_matched[j];
        return num_matched[T.length()];
    }
};
